src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,787 @@
+{
+   Initialize();
+}
+
+
+PluginPath VST3Effect::GetPath()
+{
+   return VST3Utils::MakePluginPathString( { mModule->getPath() }, mEffectClassInfo.ID().toString());
+}
+
+ComponentInterfaceSymbol VST3Effect::GetSymbol()
+{
+   return wxString { mEffectClassInfo.name() };
+}
+
+VendorSymbol VST3Effect::GetVendor()
+{
+   return wxString { mEffectClassInfo.vendor() };
+}
+
+wxString VST3Effect::GetVersion()
+{
+   return mEffectClassInfo.version();
+}
+
+TranslatableString VST3Effect::GetDescription()
+{
+   //i18n-hint VST3 effect description string
+   return XO("SubCategories: %s").Format( mEffectClassInfo.subCategoriesString() );
+}
+
+EffectType VST3Effect::GetType()
+{
+   using namespace Steinberg::Vst::PlugType;
+   const auto& cats = mEffectClassInfo.subCategories();
+   if(std::find(cats.begin(), cats.end(), kFxGenerator) != cats.end())
+      return EffectTypeGenerate;
+   if(std::find(cats.begin(), cats.end(), kAnalyzer) != cats.end())
+      return EffectTypeAnalyze;
+   
+   return EffectTypeProcess;
+}
+
+EffectFamilySymbol VST3Effect::GetFamily()
+{
+   return VST3EffectsModule::PluginType;
+}
+
+bool VST3Effect::IsInteractive()
+{
+   return mEditController != nullptr;
+}
+
+bool VST3Effect::IsDefault()
+{
+   return false;
+}
+
+bool VST3Effect::IsLegacy()
+{
+   return false;
+}
+
+bool VST3Effect::SupportsRealtime()
+{
+   return true;
+}
+
+bool VST3Effect::SupportsAutomation()
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         if(parameterInfo.flags & Vst::ParameterInfo::kCanAutomate)
+            return true;
+      }
+   }
+
+   return false;
+}
+
+bool VST3Effect::GetAutomationParameters(CommandParameters& parms)
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         parms.Write(
+            std::to_string(parameterInfo.id),
+            mEditController->getParamNormalized(parameterInfo.id)
+         );
+      }
+   }
+
+   return true;
+}
+
+bool VST3Effect::SetAutomationParameters(CommandParameters& parms)
+{
+   using namespace Steinberg;
+
+   if(mComponentHandler== nullptr)
+      return false;
+
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) != kResultOk)
+         continue;
+
+      Vst::ParamValue value;
+      if(parms.Read(std::to_string(parameterInfo.id), &value))
+      {
+         mComponentHandler->beginEdit(parameterInfo.id);
```

Comment:
beginEdit/EndEdit pair again.  Maybe make a little "ParamEditScope" utility object to do them in ctor and dtor.